examples/trivial_discord_bot.py

```diff
@@ -0,0 +1,148 @@
+    """
+
+    @staticmethod
+    def consumes_inputs() -> Set[Type[InputEvent]]:
+        return {
+            DiscordInputEvent,
+        }
+
+    def matches(self, event: InputEvent) -> bool:
+        if not isinstance(event, DiscordInputEvent):
+            return False
+
+        return event.text == "!hello"
+
+
+class DiscordTrivialCondition(Condition):
+    """
+    Allows every input.
+    """
+
+    @staticmethod
+    def consumes_inputs() -> Set[Type[InputEvent]]:
+        return {
+            DiscordInputEvent,
+        }
+
+    def allows(self, event: InputEvent) -> bool:
+        return True
+
+
+class DiscordPrintAndReplyAction(Action):
+    """
+    Print every InputEvent.
+    """
+
+    _queue: OutputQueue
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.logger = logging.getLogger(__name__ + type(self).__name__)
+
+    @staticmethod
+    def consumes_inputs() -> Set[Type[InputEvent]]:
+        return {
+            DiscordInputEvent,
+        }
+
+    @staticmethod
+    def produces_outputs() -> Set[Type[OutputEvent]]:
+        return {
+            DiscordOutputEvent,
+        }
+
+    async def act(self, event: InputEvent, state: Dict[str, Any]) -> None:
+        """
+        Construct a DiscordOutputEvent with the result of performing the calculation.
+        """
+        if not isinstance(event, DiscordInputEvent):
+            self.logger.warning("Received wrong event type %s", type(event))
+            return
+
+        self.logger.info("event.txt = %s", event.text)
+        test_event = DiscordOutputEvent(
+            text="world", message=event.message, use_message_channel=True
+        )
+
+        self.logger.info(test_event)
+        await self.send(test_event)
+
+    async def send(self, event: OutputEvent) -> None:
+        await self._queue.put(event)
+
+
+def load_secrets() -> Any:
+    secrets_file = os.path.join(__location__, "TrivialDiscordBot", "secrets.json")
+
+    # Check to see if the file exists and complain if it does not
+    if not os.path.isfile(secrets_file):
+        raise RuntimeError(
+            f"{secrets_file} did not exist - and we need secrets! (In particular, a token)"
+        )
+
+    with open(secrets_file, "rb") as json_file:
+        return json.load(json_file)
+
+
+if __name__ == "__main__":
+
+    # Acquire token from json file in the root of this dir
+    secrets = load_secrets()
+    token = secrets["token"]
+
+    # For the moment just printing all discord input events
+
+    # - Behavior
+    # Construct the print_behavior - which just prints every input event
+    print_behavior = Behaviour()
+    # Triggers on all PostInput events
+    print_behavior.add(DiscordTrivialTrigger())
+    # Passes all incoming events
+    print_behavior.add(DiscordTrivialCondition())
+    # Prints all the events which make their way into the bot
+    print_behavior.add(DiscordPrintAndReplyAction())
+
+    # - IOConfig
+    post_ioconfig = PostIOConfig()
```

Comment:
Future Thoughts/Shortcuts:If PostIOConfig is required for configuring Discord, do we want to make a system that automatically does that pipeline?
Or do we only ever do one add_io_config to bot, and so post_ioconfig should contain everything?

Reply:
Neither, really.
Building on both https://github.com/mewler/mewbot/tree/2990b637184105bdba773f0cff50007f968632f7/design-docs#input and the principle of configuration of the bot not being driven by python code, the solution is that the `mewbot.loader` will be able to load IOConfig objects of arbitary classes as needed, and those classes will automatically output the Input/Outputs that they support. (The method of building it up here is essentially the wrong way around)

Reply:
https://github.com/mewler/mewbot/pull/4/commits/c8c08db127b13daa3d56f1c157a72cbbc2468fe4 is a good example of how it is envisioned to work!